lib/hobbes/lang/preds/dependent.C

```diff
@@ -0,0 +1,588 @@
+    }
+    return TAbs::make(s, typeExprToMonoTypePtr(tabs.second));
+  }
+};
+
+struct _TApp {
+  typedef std::pair<Type*, array<Type*>*> type;
+  static MonoTypePtr build(_TApp::type& tapp) {
+    MonoTypes ms;
+    for(unsigned int i = 0; i < tapp.second->size; ++i) {
+      ms.push_back(typeExprToMonoTypePtr(tapp.second->data[i]));
+    }
+    return TApp::make(typeExprToMonoTypePtr(tapp.first), ms);
+  }
+};
+
+struct _Func {
+  typedef std::pair<array<Type*>*, Type*> type;
+  static MonoTypePtr build(_Func::type& f) {
+    Record::Members mems;
+    for(unsigned int i = 0; i < f.first->size; ++i) {
+      mems.push_back(Record::Member(".f" + str::from(i), typeExprToMonoTypePtr(f.first->data[i])));
+    }
+    return Func::make(Record::make(mems), typeExprToMonoTypePtr(f.second));
+  }
+};
+
+struct _Array {
+  typedef Type* type;
+  static MonoTypePtr build(_Array::type& ar) {
+    return Array::make(typeExprToMonoTypePtr(ar));
+  }
+};
+
+struct _FixedArray {
+  typedef std::pair<Type*, Type*> type;
+  static MonoTypePtr build(_FixedArray::type& far) {
+    return FixedArray::make(typeExprToMonoTypePtr(far.first), typeExprToMonoTypePtr(far.second));
+  }
+};
+
+struct _TString {
+  typedef array<char>* type;
+  static MonoTypePtr build(_TString::type& str) {
+    return TString::make(std::string(str->data, str->data + str->size));
+  }
+};
+
+struct _TLong {
+  typedef long type;
+  static MonoTypePtr build(_TLong::type& l) {
+    return TLong::make(l);
+  }
+};
+
+struct _TVar {
+  typedef array<char>* type;
+  static MonoTypePtr build(_TString::type& str) {
+    return TVar::make(std::string(str->data, str->data + str->size));
+  }
+};
+
+struct Type {
+  typedef variant<_Prim::type, _Record::type, _Variant::type, _OpaquePtr::type, _Exists::type, _Recursive::type, _TAbs::type, _TApp::type, _Func::type, _Array::type, _FixedArray::type, _TString, _TLong, _TVar> def;
+  def data;
+};
+
+#define BUILD(x) x ::build(*reinterpret_cast< x ::type*>(t->data.unsafePayload()))
+MonoTypePtr typeExprToMonoTypePtr(Type* t){
+  switch (t->data.unsafeTag()) {
+    case 0:
+      return BUILD(_Prim);
+    case 1:
+      return BUILD(_Record);
+    case 2:
+      return BUILD(_Variant);
+    case 3:
+      return BUILD(_OpaquePtr);
+    case 4:
+      return BUILD(_Exists);
+    case 5:
+      return BUILD(_Recursive);
+    case 6:
+      return BUILD(_TAbs);
+    case 7:
+      return BUILD(_TApp);
+    case 8:
+      return BUILD(_Func);
+    case 9:
+      return BUILD(_Array);
+    case 10:
+      return BUILD(_FixedArray);
+    case 11:
+      return BUILD(_TString);
+    case 12:
+      return BUILD(_TLong);
+    case 13:
+      return BUILD(_TVar);
+  }
+  return MonoTypePtr();
+}
+
+
+MonoTypePtr type_; // Filthy hack since I can't seem to get type def from typedb
+
+struct liftType {
+  static MonoTypePtr type(typedb&) {
+    return type_;
+  }
+};
+
+template <>
+  struct lift<Type*, false> : public liftType { };
```

Comment:
Why does this go through this `type_` global variable?  Were you avoiding explicitly constructing the `Type` structure here?

Reply:
Extreme DRY adherence, sorry. The hobbes for is much easier to refer back to, so I want it there, but getting that instance here is tricky

Reply:
I'll see if I can find a good way to do that without requiring you to repeat yourself (which would be very bad, I agree), but also without requiring this reacharound with a global variable.

Reply:
I put in a change to avoid the global variable in this case, hope you don't mind.

Reply:
Ah, that's how it's done, I was sure I'd tried that, but obviously not.